### PR TITLE
Displaying and managing notifications on the frontend

### PR DESCRIPTION
**Pull request by kamalq97**
Changelog:
* Added notification views and urls
* Moved JavaScript from **base.html** to **base.js**
* Added JavaScript functions that use Ajax to make requests to the server for deleting notifications or marking them as "read"
* Added Bootstrap dropdown menu for displaying notifications to **base.html**
* Added unit tests for notification views in **test_views.py**

### CODE REVIEW

**Comment by kamalq97** [comment]
@murradkh I have just added unit tests for notification views and merged the most recent version of the `user_notifications` branch into this branch. I can confirm the notifications created by the rules show up on the dropdown menu and direct the user to the relevant `single_stock` page 👍

**Comment by murradkh** [comment]
> @murradkh I have just added unit tests for notification views and merged the most recent version of the `user_notifications` branch into this branch. I can confirm the notifications created by the rules show up on the dropdown menu and direct the user to the relevant `single_stock` page 👍

Great! 
now left to enable the user to set their own rules configuration right?

File: myapp/urls.py [hunk 1]
```diff
@@ -8,13 +8,19 @@
...
+    path('accounts/login/', auth_views.LoginView.as_view(template_name='login.html'), name='login'),
+    path('accounts/logout/', views.logout_view, name='logout'),
+    path('accounts/watchlist/', views.watchlist_view, name='watchlist'),
     path('accounts/profile/', views.profile_view, name='profile'),
     path('accounts/password/', views.password_change_view, name='password_change'),
     path('accounts/register/', views.register, name='register'),
+    path('notifications/', views.list_notifications_view, name='list_notifications'),
+    path('notifications/unread_count/', views.notification_unread_count_view, name='notification_unread_count'),
+    path('notifications/nread/', views.notifications_mark_read_view, name='notifications_all_read'),
+    path('notifications/<int:pk>/nread/', views.notifications_mark_read_view, name='single_notification_read'),
```
Comment (murradkh) [inline myapp/urls.py:23]: if you like you can combine these two paths into one path using regex (and to have fewer views)

File: myapp/static/js/myapp/base.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    var url = myapp.URLS.notificationsUnreadCount;
+    xhr.open("GET", url, true);
+    xhr.onreadystatechange = function () {
+        if (xhr.readyState === XMLHttpRequest.DONE) {
+            var status = xhr.status;
+            if (status === 0 || (status >= 200 && status < 400)) {
+                var responseDict = JSON.parse(xhr.responseText);
+                var numNotifications = responseDict['unread_count']
+                if (numNotifications > 0){      
+                    badge.appendChild(document.createTextNode(numNotifications));
```
Comment (murradkh) [inline myapp/static/js/myapp/base.js:31]: I like it, showing the number of unread messages

File: myapp/static/js/myapp/base.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    $(this).find('.dropdown-menu').first().stop(true, true).slideDown(100);
+});
+
+
+$('.dropdown').on('hide.bs.dropdown', function() {
+    $(this).find('.dropdown-menu').first().stop(true, true).slideUp(100);
+});
+
+
+window.setInterval(getUnreadCount, 30000);
```
Comment (murradkh) [inline myapp/static/js/myapp/base.js:192]: good one, that's an alternative solution for the real-time connection between the browser and the server

**Review by murradkh — APPROVED** [review]
WAW, really amazing, all respect kamal, i'm very excited to see how our work would work together. team-work is really nice.

**Pull request merged**